### ISSUE TITLE
UI design for URI detection that allows for multiple relevant entries and just 1 relevant entry

### ISSUE THREAD

**Issue by sandysage** [mozilla-lockbox/lockbox-extension#443]
In situations where a user goes to a website address that matches a corresponding entry in their lockbox, we should allow quick access to that entry's details, specifically the ability to copy and paste username and password fields. 

**Also for Consideration**
[ ] - situations where there is more than 1 entry associated with the website address 
[ ] - URI pre-population of website address and/or name as part of the entry creation process (may need to be a separate issue)

**Comment by changecourse** [comment]
@linuxwolf what sort of flexibility is possible for exact matching vs relative matching of the website address? Im curious about when the URL changes slightly, but still maintains the same root... Happy to provide better context if this doesn't make sense.

**Comment by linuxwolf** [comment]
@changecourse some extra information may be needed.  Specifically, what constitutes a "slight change" to the URL.  What are we using for a definition of a URL (the standard, or what most people think of which (I assume) is synonymous with hostname)?

There is some flexibility in matching, but we need to be careful.  For example, it's often safe to "strip off" or ignore **a** subdomain (e.g., "youface.example" can match "m.youface.example" or "www.youface.example") -- to a point.  For example, it's usually safe to match "m.youface.example" to an existing entry for "youface.example", but _not_ to go down to just "example".  There is a regularly updated [list of public suffixes](https://publicsuffix.org/) that helps to determine where it's unsafe to ignore past.

It's best for one side of the match to stay immutable, but that would require a clean starting point that I'm not sure we have.

**Comment by changecourse** [comment]
Here's my first pass on a design worth discussion for this story... will want to discuss with the team some use cases and expectations based on the functionality and user expectations. It would be my hope that we'd want to submit this sort of concept for potential testing prior to implementation as well. 

![443_rd1](https://user-images.githubusercontent.com/6164801/35306791-dc587e10-005c-11e8-937b-5b9fe3ef7b5b.png)

My expectations for success include a user's ability to...
- understand why they are being presented with an entry 'quick view' within the context of the doorhanger
- be able to access entry details quickly
- be able to copy username (without having to dig into entry details)
- be able to copy password (without having to dig into entry details)
- have quick access to ALL entries associated with the current root domain (in progress, though not shown in screenshot above)

@hmcgaw & @sandysage  let's plan on discussing next steps for this issue.

**Comment by hmcgaw** [comment]
Hey Ryan, this looks good so far as a direction to expand on. 

Would like to see:
- What happens when the chevron next to the site name is A. hovered on B. selected
- What's the user-feedback when they select the copy button
- Hover states on username, password, and copy icons
- When search is being typed into

**Comment by changecourse** [comment]
Based on the conversations and concerns raised this morning, here is a more basic approach to URI detection, including a basic flow of potential use cases, and the corresponding doorhangers based on those options. 

![uri detection - page 1](https://user-images.githubusercontent.com/6164801/35416887-2c45385a-01e8-11e8-8647-6057365314e6.png)

![1_443](https://user-images.githubusercontent.com/6164801/35416889-3035f8aa-01e8-11e8-80b4-57e75e9e2514.png)
![2_443](https://user-images.githubusercontent.com/6164801/35416890-304e40a4-01e8-11e8-92de-90e8a4fa7d28.png)
![3_443](https://user-images.githubusercontent.com/6164801/35416892-30741130-01e8-11e8-82e0-14d527a9b5f2.png)

**Comment by changecourse** [comment]
Hello all, I've updated the approach throughout to align with our discussion around functionality of pre-populating the search, based on URI detection. Included below are the 4 different use cases depending on how many matches/entries there are.

1. No entries found for this site – display entry list with informational message status
![1 no matching entries](https://user-images.githubusercontent.com/6164801/35636290-efc1a258-066d-11e8-8b34-cedd3eaff44e.png)

2. One Entry found for this site – display the entry card in a pre-populated search list, with available quick actions for copying username and password
![2 one matching entry](https://user-images.githubusercontent.com/6164801/35636397-3a619322-066e-11e8-931c-116e527cefaa.png)

3. Two Entries found for this site – same as above, we display the entry cards in a pre-populated search list, with available quick actions for copying username and password
![3 two matching entries](https://user-images.githubusercontent.com/6164801/35636429-54f2d75a-066e-11e8-9e36-19b19651761a.png)

4. More than two entries found for this site – we display the entry cards in a pre-populated search list, WITHOUT available quick actions, due to clutter and visual noise, not to mention the vertical real-estate required, which would push some entries down out of view
![4 2 matching entries](https://user-images.githubusercontent.com/6164801/35636514-9b7ae2da-066e-11e8-9beb-2e31a50a24c7.png)

Please reach out if you'd like to discuss in more detail. 

@kimberlythegeek let me know if you'd like to touch base on any a11y thoughts as well

**Comment by kimberlythegeek** [comment]
@changecourse As far as visual design is concerned, the main considerations would be sufficient color contrast and visibly apparent focus.

Here's a [list of various contrast checkers](http://www.webaxe.org/color-contrast-tools/).

I can also review accessibility as needed when new visual interfaces/features are added.

**Comment by sandysage** [comment]
@kimberlythegeek This is awesome. Thanks for reviewing and providing great resources. From what is shown in the comps are you thinking the color contrast is sufficient or are adjustments needed?

@changecourse I think there is still an open question in #498 that could have implications to these designs: "do we want to distinguish between user-entered and the implicit?" I'd like to get that addressed before we consider this done.

Outside of that, the comps align to my expectation for acceptance in #317 including considerations for the various use cases.

**Comment by kimberlythegeek** [comment]
@sandysage Apologies for the delay! Looking at it, most of the contrast seems sufficient. Typically the tools for color contrast require the hex values of the text and background colors. There are browser-based tools, but they won't check contrast within an image.

**Comment by kimberlythegeek** [comment]
@sandysage Apologies for the delay! Looking at it, most of the contrast seems sufficient. Typically the tools for color contrast require the hex values of the text and background colors. There are browser-based tools, but they won't check contrast within an image.

**Comment by kimberlythegeek** [comment]
Talked to Yura about his preferred method for testing mockups for sufficient color contrast. He suggested [this desktop application](https://developer.paciellogroup.com/resources/). 
He said he usually uses the color picker 
11:44:51 AM <yzen> yeah :( that's mostly how i did that too get color picker, pick colors, test, but i think you should have better chance with https://developer.paciellogroup.com/resources/contrastanalyser/

**Comment by kimberlythegeek** [comment]
Talked to Yura about his preferred method for testing mockups for sufficient color contrast. He said he usually uses the color picker, gets the value, and tests using a contrast checker.

He suggested [this desktop application](https://developer.paciellogroup.com/resources/) by The Paciello Group.

**Comment by m8ttyB** [comment]
tl;dr, what @changecourse has mocked out looks good. PI will validate the UX once it has it has gone through development.
